DM-3690

jonathansick opened the pull request.

jonathansick committed:
Create package structure

- Create setup.py, use setuptools
- Add gitignore
- Add COPYRIGHT file (RFC-45 syle)

jonathansick committed:
Representations of Category and Topic JSON feeds

Retrieve data from the Discourse API through JSON feeds. We use requests
to obtain the feeds, which also automatically parses the JSON for us.

Properties give easy access to results from the API.

jonathansick committed:
Create a TopicCache

So that duplicate topic notification emails aren't repeatedly sent, the
TopicCache serves as an interface to a simple JSON document describing
the topics that have been sent.

Once an email is sent the corresponding topic is added to the cache. At
the end of all processing the cache is flushed to disk (as JSON).

TopicCache implements a dict-like API so it's easy to do `in`-type
membership testing of topics.

jonathansick committed:
Create forward_discourse script

This script takes a JSON file mapping category names to lists of email
addresses and forwards new topics in those categories to the appropriate
email addresses. Mandrill is used to send transactional emails using an
HTML template.

jonathansick committed:
Add new_topics method to CategoryFeed

This method compares topics in a category to the cache to provide a list
of topics that should be emailed out.

jonathansick committed:
config.json to hold our category->email config

This JSON file maps category_id to the recipients dict used by the
Mandrill API.

Currently setup for testing only. Eventually this file will be the
configuration used in operations.

jonathansick committed:
Add discourse_categories script

This script prints out the category ID and the name. The category id
(an int) is used in the forward_discourse script's JSON input file
to reliably identify categories.

jonathansick committed:
Add Getting Started guide to README

jonathansick committed:
Add SiteFeed to interface with site.json

site.json is useful for getting a listing of all categories on the site.

jonathansick committed:
Use SiteFeed to get a category listing

jonathansick committed:
Include category name in subject line

This should provide better context for the source of the email and help
those who are filtering messages.

I may revisit this decision when emails are sent from mailing lists that
add their own information to the subject line.

jonathansick committed:
Update template to community-mailbot-template-2

- Include topic author
- Template had better footer and fixes link in the call-to-action
  button

jonathansick committed:
Cleanup links in Discourse's HTML

Discourse's baked HTML includes both relative URLs (`/users/jsick`) and
schema-agnostic URLs (`//community.lsst.org`). This commit adds a
clean_discourse_html() function, run by the forward_discourse script, that
processes all <a> and <img> tags and ensures all URLs are absolute and
fully specified.

BeautifulSoup4 is used to do search-and-replace in the HTML document.
The urllib.parse standard module provides robust URL splitting and
joining.

Includes unit tests for making absolute URLs and instructions on reading
the tests.

This adds BeautifulSoup4 as a dependency.

jonathansick committed:
Better url building with urlsparse/urlunsplit

The feeds now contruct their URLs from urllib.parse.urlunsplit, which
constructs URLs from the types of len-six tuples made by urlparse.
This could be better since scheme, domain and path are more explicitly
specific.

Under this scheme, the base forum URL should either provide a URL or at
least `//domain.tld` to be scheme agnostic. Providing a base url like
`community.lsst.org` to the forward_discourse script will look like the
URL path, not domain, is `community.lsst.org`.

jonathansick committed:
Correct license to MIT

jonathansick committed:
Fix README

- address for pip install via github
- add reference configuration information

jonathansick committed:
Add __init__.py for scripts module

This is needed for setuptools entrypoints given a regular pip install.

jonathansick committed:
Fix forward_discourse --cache default value

It was missing in the original ArgumentParser configuration.

jonathansick committed:
Remove Unicode from print

The Amazon Linux AMI I'm running on for production doesn't like printing
out unicode characters. Removing it for now.

jonathansick committed:
Fix bug in checking existance of cache's dir

jonathansick committed:
Add DM Notifications to default/test config

jonathansick committed:
Use slugs to get paths to category feeds

Adds a SiteFeed.category_paths property that maps category ID to path.

It turns out that URL paths to category feeds based on their slugs is
better than using category IDs since subcategories aren't available at
their integer ID path.

e.g. community.lsst.org/c/10.json works,
but  community.lsst.org/c/10.json doesn't because 11 is a subcategory.

However, http://community.lsst.org/c/dm/dm-notifications does work.

This is implemented in CategoryFeed and the forward_discourse script.

jonathansick committed:
Explicitly get only the Latest feed for categories

jonathansick committed:
Use a post's title, not fancy_title

fancy_title uses HTML escaping that we don't want for emails

jonathansick committed:
Save cache after each message for reliability

Even if forward_discourse crashes mid-run, we want all sent messages to
be cached to avoid repeat messages.

jonathansick committed:
Use requests.get params to insert Discourse auth

I did try adding them with urlunsplit, but it really is better to let
requests do this.

jonathansick committed:
Fix authentication to private topics

Topics feeds can be accessed with multiple URLs:

1. community.lsst.org/t/topic-slug.json
2. community.lsst.org/t/topic-slug/topic-id.json
3. community.lsst.org/t/topic-id.json

Previously I used #1, but this was incompatible with Discourse auth;
the URL was being redircted to #2 and losing the auth params. So instead
we use the simplest form that doesn't redirect, #3.

TopicFeed still needs the slug to supply the HTML URL. Thus
CategoryFeed.new_topics now supplies a list of named tuples that
includes attributes with both the ID and slug of a topic.

jonathansick committed:
Set the default_user argument

jonathansick committed:
Compete documentation with how to setup cron task

ktlim commented on README.rst:
```diff
@@ -2,11 +2,141 @@
-The ``community_mailbot`` is a friendly bot that tells subscriber's to LSST DM's legacy Mailman email lists about things happening on `community.lsst.org <http://community.lsst.org>`_.
+The ``community_mailbot`` is a friendly bot that tells subscribers to LSST DM's legacy Mailman email lists about things happening on `community.lsst.org <http://community.lsst.org>`_.
+
+The bot likes to be awoken regularly by ``cron``; each time it will ask the Discourse server about latest messages in different categories that the bot tracks (you can set this with a simple ``config.json`` file).
```
"awakened"

ktlim commented on README.rst:
```diff
@@ -2,11 +2,141 @@
+The ``community_mailbot`` is a friendly bot that tells subscribers to LSST DM's legacy Mailman email lists about things happening on `community.lsst.org <http://community.lsst.org>`_.
+
+The bot likes to be awoken regularly by ``cron``; each time it will ask the Discourse server about latest messages in different categories that the bot tracks (you can set this with a simple ``config.json`` file).
+If there is are new topics, ``community_mailbot`` will send an email the appropriate email lists using its buddy, `Mandrill <http://mandrillapp.com>`_.
```
Delete "is", add "to" after "send an email"

ktlim commented on README.rst:
```diff
@@ -2,11 +2,141 @@
+   source activate community_mailbot
+   forward_discourse /home/ec2-user/config.json
+
+   echo "$(date) Ran forward_discourse" > /home/ec2-user/forward_discourse.log
```
Do you want ">>" here? Is this really necessary, since it's not really more informative than the cron log?

ktlim commented on community_mailbot/scripts/forward_discourse.py:
```diff
@@ -0,0 +1,203 @@
+    parser.add_argument(
+        '--key',
+        default=default_discourse_key,
+        help='Discourse API key ({0})'.format(default_discourse_key))
```
It's not good practice to print the (secret) key to the screen.

jonathansick replied:
I guess a good compromise would just be to say that the environment variable can be used in lieu of explicitly setting `--key` etc?

ktlim replied:
I think that would be fine.

ktlim commented on community_mailbot/scripts/forward_discourse.py:
```diff
@@ -0,0 +1,203 @@
+    parser.add_argument(
+        '--mandrill',
+        default=default_mandrill_key,
+        help='MANDRILL API key ({0})'.format(default_mandrill_key))
```
Same here about printing keys.

ktlim commented on README.rst:
```diff
@@ -2,11 +2,141 @@
+   {
+       "<id>": [{"email": "<email address>",
+                 "name": "<recipient name>",
+                 "type": "to"}],
```
Clarification about the "type" (which is a Mandrill thing, I guess) would be good -- always say "to"? are there other possible values?

timj commented on setup.py:
```diff
@@ -0,0 +1,58 @@
+    full_filename = os.path.join(
+        os.path.abspath(os.path.dirname(__file__)),
+        filename)
+    return open(full_filename).read()
```
Should this `open` take an encoding? I know this is just a `setup.py` but ideally I'd like the file to be closed when we have done reading it (with a `with`).

timj replied:
It has been pointed out to me that the file does get closed on function exit.

jmatt commented:
I reviewed and discussed the code and architecture with @jonathansick yesterday. Here is a list some of the top ideas that came out of the discussion. It generally looks good and only small changes are suggested.

#### TODOs
- Add an optional new argument to display secrets and configuration in a meaningful way.
- Remove displaying secrets in the help descriptions (per @ktlim's comments earlier).
- Add a simple configuration module usable from the scripts or library.
- Add optional argument to pass in a specific configuration file. This will help if the method of deployment changes. IE Docker, different distro, new best practices, etc.
- Move from a mix of environmental variables and configuration file to just a configuration file.
- Add configuration file to .gitignore. Keep example or template of the configuration file in git. This will prevent any accidental check-in of secrets.
- Move away from using a shell wrapper in the cron job and use the virtualenv python script and arguments instead. This is more portable and transparent in the long term.

#### Discussed
- There is no commonly used python API for discourse. That's why this module uses requests directly against discourse instead of some existing discourse API.
- Using JSON for config now, which is unusual for python but not unheard of. This is sufficient for the service's requirements. That's fine since there are no standards around how configuration should work. No reason to use ConfigParser, etc instead of JSON.
- It's possible to improve the structure of the scripts. But it's not necessary since it's small and unlikely to be reused. Discussed keeping all the logic and rules in the library and the script only does data manipulation, validation, teeing up so it can use the library.
- Similarly, It's possible to improve the structure of the discourse module. But it's not necessary. Top ideas discussed: Classes could benefit from inheritance, could be hard to test due to some side effects and no functional way to pass in that that. But the library is small and the fixes are simple. So no additional action required.

ktlim commented:
Once upon a time we were trying to promulgate a standard of YAML for configuration.  It hasn't caught on widely, but I'd still like to suggest it.

jonathansick committed:
Minor README fixes

Typos and clarification on the structure of the Mandrill API signature
in the README given code review.

For DM-3690.

jonathansick committed:
Remove secret key ENV variable contents from help

Resolves a security concern. Ultimately all secret keys should be stored
in the configuration file only.

For DM-3690.
